src/v/cluster/scheduling/types.h

```diff
@@ -230,7 +229,24 @@ class allocated_partition {
 
 private:
     friend class partition_allocator;
-    void add_replica(model::broker_shard, allocation_state&);
+
+    // construct an object from an original assignment
+    allocated_partition(
+      std::vector<model::broker_shard>,
+      partition_allocation_domain,
+      allocation_state&);
+
+    struct previous_replica {
+        model::broker_shard bs;
+        size_t idx;
+        std::optional<model::broker_shard> original;
+    };
+
+    std::optional<previous_replica> prepare_move(model::node_id previous);
```

Comment:
Was wondering if we should wrap the return type of this into a RAII object and call 'cancel_move' in the destructor unless the move is committed..sort of like allocation_units.. that avoids `ss::defer` style code in the callers.

handle = partition.prepare_move(...)..
....
handle.commit();

Reply:
These are private helpers so I didn't want to bother with setting up destructors and stuff. IMO code with defer is quite readable.